### PR TITLE
New VFS API

### PR DESCRIPTION
**Pull request by matklad**
cc @flodiebold , @jonas-schievink

### CODE REVIEW

**Comment by matklad** [comment]
The branch that integrates changes into rust-analyzer is https://github.com/matklad/rust-analyzer/tree/vfs

File: crates/vfs/src/loader.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    pub fn local_cargo_package(base: AbsPathBuf) -> Entry {
+        Entry::Directory { path: base, globs: globs(&["*.rs", "!/target/"]) }
+    }
+    pub fn cargo_package_dependency(base: AbsPathBuf) -> Entry {
+        Entry::Directory { path: base, globs: globs(&["*.rs", "!/tests/", "!/examples/"]) }
+    }
+}
+
+fn globs(globs: &[&str]) -> Vec<String> {
+    globs.iter().copied().map(|it| it.to_string()).collect()
```
Comment (lnicola) [inline crates/vfs/src/loader.rs:46]:
```suggestion
    globs.iter().map(|it| it.to_string()).collect()
```

File: crates/vfs/src/loader.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+impl Entry {
+    pub fn rs_files_recursively(base: AbsPathBuf) -> Entry {
+        Entry::Directory { path: base, globs: globs(&["*.rs"]) }
+    }
+    pub fn local_cargo_package(base: AbsPathBuf) -> Entry {
+        Entry::Directory { path: base, globs: globs(&["*.rs", "!/target/"]) }
+    }
+    pub fn cargo_package_dependency(base: AbsPathBuf) -> Entry {
+        Entry::Directory { path: base, globs: globs(&["*.rs", "!/tests/", "!/examples/"]) }
```
Comment (jonas-schievink) [inline crates/vfs/src/loader.rs:41]:
```suggestion
        Entry::Directory { path: base, globs: globs(&["*.rs", "!/tests/", "!/examples/", "!/benches/"]) }
```

File: crates/vfs/src/walkdir_loader.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+use globset::{Glob, GlobSetBuilder};
+use paths::{AbsPath, AbsPathBuf};
+use walkdir::WalkDir;
+
+use crate::loader;
+
+#[derive(Debug)]
+pub struct WalkdirLoaderHandle {
+    sender: crossbeam_channel::Sender<Message>,
+    _thread: jod_thread::JoinHandle,
```
Comment (jonas-schievink) [inline crates/vfs/src/walkdir_loader.rs:14]: The order of this is significant, right? Maybe add a comment?

File: crates/vfs/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+                (file_id, path)
+            })
+    }
+    pub fn set_file_contents(&mut self, path: VfsPath, contents: Option<Vec<u8>>) {
+        let file_id = self.alloc_file_id(path);
+        let change_kind = match (&self.get(file_id), &contents) {
+            (None, Some(_)) => ChangeKind::Create,
+            (Some(_), Some(_)) => ChangeKind::Modify,
+            (Some(_), None) => ChangeKind::Delete,
+            (None, None) => ChangeKind::Delete,
```
[the remark this answers is not in the document]
Comment (matklad) [reply crates/vfs/src/lib.rs:106]: Hm, good point, if we register a change we *need* `Delete` (as that's the only one that would return `false` from `exists`), but naturally we can just skip the change

File: crates/vfs/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+//! A simple `WalkdirLoaderHandle` is provided, which doesn't implement watching
+//! and just scans the directory using walkdir.
+//!
+//! VFS stores a flat list of files. `FileSet` can partition this list of files
+//! into disjoint sets of files. Traversal-like operations (including getting
+//! the neighbor file by the relative path) are handled by the `FileSet`.
+//! `FileSet`s are also pushed to salsa and cause it to re-check `mod foo;`
+//! declarations when files are created or deleted.
+//!
+//! `file_set::FileSet` and `loader::Entry` play similar, but different  roles.
```
Comment (bjorn3) [inline crates/vfs/src/lib.rs:30]:
```suggestion
//! `file_set::FileSet` and `loader::Entry` play similar, but different roles.
```

**Comment by matklad** [comment]
Let's land it, it's unused yet, but it should be easier to base further work on top. 

bors r+

**Pull request merged**